mifosng-provider/src/main/java/org/mifosplatform/portfolio/loanaccount/service/LoanSchedularServiceImpl.java

```diff
@@ -97,9 +101,46 @@ public void applyChargeForOverdueLoans() throws JobExecutionException {
     @Override
     @CronTarget(jobName = JobName.RECALCULATE_INTEREST_FOR_LOAN)
     public void recalculateInterest() {
+    	Integer maxNumberOfRetries = ThreadLocalContextUtil.getTenant().getConnection().getMaxRetriesOnDeadlock();
+    	Integer maxIntervalBetweenRetries = ThreadLocalContextUtil.getTenant().getConnection().getMaxIntervalBetweenRetries();
         Collection<Long> loanIds = this.loanReadPlatformService.fetchLoansForInterestRecalculation();
+        int i = 0;
         for (Long loanId : loanIds) {
-            this.loanWritePlatformService.recalculateInterest(loanId);
+        	logger.info("Loan ID " + loanId);
+        	Integer numberOfRetries = 0;
+        	while (numberOfRetries <= maxNumberOfRetries) {
+        		try {
+        			this.loanWritePlatformService.recalculateInterest(loanId);
+        			numberOfRetries = maxNumberOfRetries + 1;
+        		}catch (CannotAcquireLockException
+        				 | ObjectOptimisticLockingFailureException exception) {
+        			logger.info("Recalulate interest job has been retried  "
+        					 	+ numberOfRetries + " time(s)");
+        			/***
+        			 * Fail if the transaction has been retired for
+        			 * maxNumberOfRetries
+        			 **/
+        			 if (numberOfRetries >= maxNumberOfRetries) {
+        			 	logger.warn("Recalulate interest job has been retried for the max allowed attempts of "
+        			 			+ numberOfRetries + " and will be rolled back");
+        			 	throw (exception);
+        			 }
+        			 /***
+        			  * Else sleep for a random time (between 1 to 10 seconds)
+        			  * and continue
+        			  **/
+        			 try {
+    				 	Random random = new Random();
+    				 	int randomNum = random.nextInt(maxIntervalBetweenRetries + 1);
+    				 	Thread.sleep(1000 + (randomNum * 1000));
+    				 	numberOfRetries = numberOfRetries + 1;
+        			} catch (InterruptedException e) {
+        				 	throw (exception);
+        			}
+        		}
+        		i++;
+        	}
+        	logger.info("Loans count " + i); 
```

Comment:
Add this log to string and finally throw JobExecutionException with this string for logging purpose at job level